### ISSUE TITLE
[Discussion] Compress datasets with bzip2

### ISSUE THREAD

**Issue by BR0kEN-** [thegreenwebfoundation/admin-portal#118]
### Original size

```
$ du -sh green_urls_2021-02-19.db
121M	green_urls_2021-02-19.db
```

### gzip

```
$ gzip --force --best green_urls_2021-02-19.db
$ du -sh green_urls_2021-02-19.db.gz
37M	green_urls_2021-02-19.db.gz
```

### bzip2

```
$ bzip2 --force --compress green_urls_2021-02-19.db
$ du -sh green_urls_2021-02-19.db.bz2
27M	green_urls_2021-02-19.db.bz
```

### Result

~10 MB smaller size of a resulting archive

### Thoughts

I'm concerned about the existing solutions (like my own) built around the currently available `gzip` archives. Its logic is the following:
- visit https://admin.thegreenwebfoundation.org/green-urls and parse dataset URLs using `/href="(.+?\.db\.gz)"/`;
- download the latest available dataset if no local file exists or check the local's dataset release date to determine if there is a newer version available.

Considering the above, suddenly replacing `gzip` archives with `bz2` will break such integrations.

I don't think that publishing `bz2` in addition to `gzip` would make sense because this almost doubles the size in S3 for a single dataset version.

**Comment by mrchrisadams** [comment]
Hi @BR0kEN- I didn't realise that bzip2 compressed it so much better than gzip!

To be honest, I have been experimenting with zstd for this, but the savings didn't seem worthwhile for the reduced support. Here's how it measured after your issue her got me thinking about formats:

```
.rw-rw-rw-  29M chrisadams  6 Mar 14:51 green_urls_2021-03-06.db.bz2
.rw-rw-rw-  32M chrisadams  6 Mar 14:51 green_urls_2021-03-06.db--level-19.zst
.rw-rw-rw-  35M chrisadams  6 Mar 14:51 green_urls_2021-03-06.db--level-10.zst
.rw-rw-rw-  36M chrisadams  6 Mar 14:51 green_urls_2021-03-06.db--level-8.zst
.rw-rw-rw-  41M chrisadams  6 Mar 14:51 green_urls_2021-03-06.db.gz
.rw-rw-rw- 140M chrisadams  6 Mar 14:51 green_urls_2021-03-06.db
```

If bzip is supported on most platforms by default, then I think it would make sense to use it - I wasn't sure if _anyone_ was using these 😄 

To be honest, I was considering changing the backup strategy later on, here so there would only be a well compressed weekly or monthly full snapshot, and then daily diffs - this isn't something I know so much about, or what would be useful to have, so I'm open to ideas, but I should be clear that right now the focus is tidying up the write-API, and getting the browser extensions refreshed.

**Comment by BR0kEN-** [comment]
I have never used `zstd`. Am I right, that to just unpack the dataset I would need to install https://github.com/facebook/zstd? That doesn't seem to be a good idea.

I'm not sure about incremental backups though. Is it possible with SQLite? Even if it is, wouldn't it bring more complexity?

In case of providing a monthly snapshot for the past 12 months plus a snapshot per week of the current month the total size can be calculated using `(SIZE MB * 12 months) + (SIZE MB * 4 weeks of the current month)`. In the case of a compressed snapshot weighting 40 MB, the consumed space would equal 640 MB (`(40*12)+(40*4)`).

Do you have any backup data size limitations or do you plan to introduce them? IMO, a couple of the latest backups should be more than enough. Let's say for the previous month + weekly for the current one.

**Comment by mrchrisadams** [comment]
# incremental backups and a strategy

My thinking for the incremental backups was based on the idea that the every entry in the green domain table has a timestamp, so it's fairly simple to get as list of updated domains, by adding a WHERE clause in the export command - something like this pseudo code:

```
SELECT * FROM green_domains WHERE modified > (YESTERDAY)
```

Constructing database, would be taking the last full snapshot then importing the daily updates to get up to the latest day.

BUT TBH, these are so small that just having the last week or month's worth of daily snapshots combined with the monthly full snapshot would likely be simple enough.

As it happens, you _can_ do incremental backups with sqlite3, [using something like litestream](https://litestream.io/), but for this case, I really think I prefer your approach.

### Compression formats

I accept your point about zstd - it's nice, but a bit of a pain.

I don't have strong preferences over the compression formats. If bzip gives 25% smaller files, and is almost as common as gzip for storage, I'm happy switching, but it would make sense to provide a bit of notice first.

If we can merge in your PR in #119, I'm happy to generate both for a month or so, and when it's come time just have the one month snapshot, we just store bz2 compressed files as the longer term archive.

**Comment by BR0kEN-** [comment]
Does it assume taking the result of `SELECT` and performing inserts based on it or what is the process of an incremental update? In MySQL-like databases, I'd use `mysql -uUSER -pPASSWD my_db < ./snapshot_2021_03_08.sql` and that merged the snapshot with an existing database. I'm not aware of the possibility to do something similar with SQLite. Is there such an option?

Anyway, the incremental update sounds like unnecessary complexity for the moment.

**Comment by mrchrisadams** [comment]
Yeah, something like that - that was what I was thinking earlier.

I understand it, you _can_ have concurrent 'users' accessing the same sqlite database as there is some concurrency control in it, meaning you could have multiple reader processes and even a writer process. [There's more here in the SQLite3 docs](https://sqlite.org/lockingv3.html). It's not something I've used before in production though.

But to be honest, after having a think given how small the files and data are in general, I think it's wiser to not have any logic at that level at all. Working with whole table exports seems much simpler.
